src/test/java/ArrayStackTest.java

```diff
@@ -1,11 +1,91 @@
+import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 
+import java.util.EmptyStackException;
+
 import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
 
 class ArrayStackTest {
 
     @Test
-    void isEmpty() {
-        assertNull(null);
+    @DisplayName("A new stack starts empty.")
+    void aNewStackIsEmpty() {
+        Stack<Integer> stack = new ArrayStack<>();
+        assertTrue(stack.isEmpty());
+    }
+
+    @Test
+    @DisplayName("An empty stack has size 0.")
+    void emptyStackHasSizeZero() {
+        Stack<Integer> stack = new ArrayStack<>();
+        assumeTrue(stack.isEmpty());
+        assertEquals(0, stack.size());
+    }
```

Comment:
Ok! I was hoping you'd come to a different conclusion than to accept _assumptions_ as being useful here.

I linked [assumptions](https://junit.org/junit5/docs/current/user-guide/#writing-tests-assumptions), but that doesn't tell you anything. There's some code which is easy to make assumptions (pun?) about, but doesn't motivate why you'd ever use an _assumption_. It does, however, link to the [javadoc](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assumptions.html).

> Assumptions are typically used whenever it does not make sense to continue execution of a given test method — **for example, if the test depends on something that does not exist in the current runtime environment.**

Emphasis mine. In my life, I've only used assumptions to skip service tests (and it was hacky for speedier in-IDE development, without standing up a container locally).

In this case, if the assumption is false (a new stack is empty), should the test be skipped, or fail? Let's clarify the difference: skipping effectively means the test succeeds, failing the test will break the build. If the new stack is non-empty, this gets skipped, and _we learn nothing about the behaviour of our code_. If we don't know the code is correct, should we be shipping it to customers, or breaking the build? I hope that answer is pretty obvious!

Funnily enough, everything I said before was true. Multiple assertions make it hard to isolate errors. You should always test the interface and not the implementation. Even assumptions isolate the errors. But, most important of all -- and what I left out -- is that the tests must correctly verify the correctness of your code. They should do it so our customers don't have to.

So. What should this test look like? Probably just `assertEquals(0, new ArrayStack<Object>().size())`. What about error isolation? Turns out, this takes practice -- in test writing and seeing tests fail (a lot). If the assumption is false, several tests will break, then it's some nerd's job to figure that out. With enough practice, you'll pretty quickly isolate the actual failure. You end up using _other tests_ as assumptions. Assuming a new stack is empty is already asserted, so best is to make use of that instead. It's also an argument for completeness in the test suite as well -- tests build on each other, especially when testing a highly stateful class like a collection.

BUT. Junit! :heart: [Nested tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-nested) do this substantially better. All the tests still run (so build would get broken), but the links between tests are explicitly shown. Isolating the error becomes going to the shallowest failure. Their example is even testing a stack.